### ISSUE TITLE
Marks with style are not apply correctly in Safari when CJK IME input.

### ISSUE THREAD

**Issue by Xheldon** [ProseMirror/prosemirror#971]
### Issue details

The editor use these prosemirror related library below:
prosemirror-model 1.7.1
prosemirror-state 1.2.3
prosemirror-transform 1.1.3
prosemirror-view 1.10.0

### Simplest code to reproduce

schema:
```
// Imports omitted
const schema = new Schema({
  nodes: {
    doc: {content: 'block+'},
    text: {group: 'inline'},
    p: {group: 'block', content: 'inline*', toDOM(node){return ['div', 0]}}
 },
  marks: {
    forecolor: {
      attrs: {
        color: {}
      },
      toDOM: mark => {
        return ['span', { style: `color: ${mark.attrs.color};` }];
      }
    }
  }
});
const state = EditorState.create({
  schema
});
let view = new EditorView(document.getElementById('root'), {
  state,
  dispatchTransaction (tr) {
    // just for log
    const newState = state.apply(tr);
    view.updateState(newState);
}
});
```
and then, i invoke the command:
```
let i = 0;
handler() {
  const {
     state,
     state: {
        tr,
        schema,
      },
      dispatch
   } = view;
   const { empty, ranges } = state.selection;
   const markType = schema.marks.forecolor;
   const applyMark = (from, to) => {
      let color = (i++ % 2) ? 'red' : 'green';
      const mark = markType.create({ color });
      if (empty) tr.addStoredMark(mark);
      else tr.addMark(from, to, mark);
      };
      ranges.forEach(range => applyMark(range.$from.pos, range.$to.pos));
      dispatch(tr);
}
```

### Setp to reproduce
 0. There have one word in editor: '啊', select it.
1. Invoke the set color function, so prosemirror will set the word '啊' to green.
2. Then, put the cursor behind the word, and Invoke the set color function, so prosemirror will set color to red in storedMark.
2. Use CJK IME(i.e. Chinese) and typing 'a'(or someelse character, it's doesn't matter)
3. Then typing Space or Enter to confirm the input, so the word '啊' will be input
7. The weird thing is showing: the two word '啊' is all be red color.

maybe my english skill is not good enough to describe the bug, so here are the gif:

![bug (1)](https://user-images.githubusercontent.com/9780154/64426510-4841c680-d0e1-11e9-80b5-f5823174a6f9.gif)

### Affected platforms

macOS Mojave 10.14.6 (18G87)
Safari 12.1.2 (14607.3.9)

**Comment by Xheldon** [comment]
Not only the color, but font-size or someelse mark that set to dom style is all broken :-(

**Comment by Xheldon** [comment]
I log the editor's child DOM count, and find the bug is related with input event and the inputType is deleteCompositionText(just safari have this type of input)

**Comment by jljorgenson18** [comment]
@Xheldon Do you have parseDOM functions set on your schema? IMEs will mutate the dom in weird ways and prosemirror often has to parse the resulting changes back into the document. Without a parseDOM function present, it would end up reparsing it as plain text.

**Comment by Xheldon** [comment]
@jljorgenson18  Sorry, but yes, i have lost some code before, here are the more details below:
```
forecolor: {
    attrs: {
         color: {}
    },
    parseDOM: [{
         style: 'color',
         getAttrs: color => {
             return color ? { color } : false;
     }],
   toDOM: mark => {
       return ['span', { style: `color: ${mark.attrs.color};` }];
   },
}
```
but it's still broken, and become more weird: after i seleted and change the color, then click the position behind it and change color, type use IME behind it, finally, when i type space or enter key to confirm the input, the previous's word's color style are disappeared, and new input word's style color is right.
and only after the span which style color that created by addMark methods has  been seleted is be influenced.
Besides that, chrome's behaviour is somewhat strange(it's result is right though), when input IME to apply a storedMarks behind a mark style in editor, the curosr's position is before first character, keeping typing the second and more character, the cursor will back to normal.

**Comment by Xheldon** [comment]
Chrome: 
![marks-bug](https://user-images.githubusercontent.com/9780154/64525828-54719200-d334-11e9-92fb-a3cbcbfb7f15.gif)
Safari:
![safari-input-bug](https://user-images.githubusercontent.com/9780154/64525849-62271780-d334-11e9-826b-7bb33ce0ed38.gif)

**Comment by jljorgenson18** [comment]
@Xheldon That is quite odd. The browser might be getting rid of the color itself and it can't be parsed back into the model. As a quick test, you could add a `data-color` attribute and add an extra parseDOM rule. It could look like this.

```js
forecolor: {
    attrs: {
         color: {}
    },
    parseDOM: [{
         style: 'color',
         getAttrs: color => {
             return color ? { color } : false;
        }
     }, {
         tag: '[data-color]',
         getAttrs: dom => {
             const color = dom.getAttribute('data-color');
             return color ? { color } : false;
        }
     }],
   toDOM: mark => {
       return ['span', { style: `color: ${mark.attrs.color};`, 'data-color': mark.attrs.color }];
   },
}
```

I would also see if this applies to every single mark or just a few. If it's just a few, then it could be this parsing issue. If not, it could be something else.


There is also a bug with deleteCompositionText on Safari where it will sometimes delete too much. It's a CSS dependent bug if your wrappers are position relative. Are you styling these spans with anything extra?

**Comment by Xheldon** [comment]
@jljorgenson18 No, there are no other styles on the span except for color style. But the data-color attribute surely applies to every mark which have color styles. And when set 'data-color' on the attribute, after the fourth setp of i mention before, the dom just become like this:
![image](https://user-images.githubusercontent.com/9780154/64590615-8806f780-d3da-11e9-926b-d33b95ca480d.png)

It's seem like that the caret-color is redundant and it's make the prosemirror getAttrs methods match dom fail?
I put the code in codepen, so you can have a try. Thx!

**Comment by Xheldon** [comment]
@marijnh  Any idea?

**Comment by jljorgenson18** [comment]
So I just double checked and this is a bug in our editor as well. Safari is mutating the dom in weird ways during compositions around color. You could try experimenting with setting an css variable instead of color to see if color is the real culprit.

```js
return ['span', { style: `--mark-color: ${mark.attrs.color};` }];
```

and then in your css

```css
span {
  color: var(--mark-color, inherit);
}
```

**Comment by Xheldon** [comment]
@jljorgenson18  I checked the code your provided and find that's surely fixed the color bug, but it's bring other problem like switch to uno and switch back then, we decide to waiting the author @marijnh to fix it or fork the library to find whether we can make a pull request to fixed it😂 Thx for help!

**Comment by marijnh** [comment]
This issue has been moved to [issue #971](https://code.haverbeke.berlin/prosemirror/prosemirror/issues/971) on the [repository's new home](https://code.haverbeke.berlin/prosemirror/prosemirror/).
